TINKERPOP-2679 update javascript driver to support stream processing

tkolanko opened the pull request:
This PR implements a change to javascript driver for allowing an optional callback which will be run with the result set of each chunk of data returned from the gremlin server rather than waiting for the entire query to finish.

docker/build.sh -t -i -n passes all tests

spmallette commented:
Thanks for submitting this - it may have to wait for 3.5.3 though as we're getting ready for release. Is that ok for you?

tkolanko commented:
@spmallette that works just fine - thanks!

jorgebay commented on docs/src/reference/gremlin-variants.asciidoc:
```diff
@@ -1721,6 +1721,32 @@ IMPORTANT: The preferred method for setting a per-request timeout for scripts is
+[source,javascript]
+----
+
+await client.submit("g.V()", {}, { batchSize: 25 }, (data) => {
```
I think mixing promises and callbacks in the same API can be confusing and prone to misuse.

I think we should expose a different method for "streaming" or grouping into smaller sets of results, for example with async iterables:

```javascript
const stream = client.stream(traversal);
for await (const item of stream) {
  statement
}
```

or regular callbacks
```javascript
client.forEach(traversal, item => {
 // called for each item
}, err => {
  // called at the end or when there's an error
});
```

jorgebay replied:
> we don't operate on traversals

I used the variable name `traversal` for the string script containing the traversal, I edited the code example for clarity.

tkolanko replied:
ah ok, let me take a look at how complex it would be to add an async iterator to websocket messages. Through a quick google search there are [some](https://github.com/alanshaw/it-ws) [libraries](https://github.com/alanshaw/it-ws) but they don't seem to be maintained

jorgebay replied:
I think from the user POV, having a thing (`ResultSet`, `Stream`, ...) that represents the result of the execution is more familiar than having 1 per "batch". The batch is a low level thing.

There are several db client libraries that use this approach (blocking iterators in other languages, async iterators in node.js, streams or callbacks). I [worked on the paging api for Cassandra](https://github.com/datastax/nodejs-driver/tree/master/doc/features/paging) that uses several of these approaches (see [here for example](https://github.com/datastax/nodejs-driver#row-streaming-and-pipes)), but these are popular patterns across node.js db client libraries (see [mysql](https://github.com/mysqljs/mysql#streaming-query-rows))

I think that we should return an object that fires events or a stream of result items, we should clearly signal when all the data has been received or when there was an error. If we use callbacks, I think we shouldn't mix it with promises to get the error.

jorgebay replied:
I think an async iterator is harder to implement in this case given that the tinkerpop protocol doesn't allow to fetch the next page (it will continue to serve the next "partial content").

I think an event emitter (emitter of result items) or a stream fit better in this case.

tkolanko replied:
@jorgebay I implemented a new `client.stream` which returns a readable stream. Can you take a look and let me know what you think.

jorgebay replied:
Nice!
I'll try to review it this week

spmallette replied:
we're getting a bit closer to 3.5.3 code freeze i think. i was wondering if there was anything left to do on this PR to have it ready for merge. cc/ @jorgebay

tkolanko replied:
Some of the existing tests don't like the changes I made to reduce duplicate code the original submit and the new stream method. I'll back out of those and leave the duplicated code for now; it's not that much. If we want to have this fixed up for 3.5.3 I can take a crack at it this weekend.

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js:
```diff
@@ -290,13 +296,31 @@ class Connection extends EventEmitter {
-        handler.result.push.apply(handler.result, response.result.data);
+        if (this._onDataMessageHandlers[response.requestId]) {
+          this._onDataMessageHandlers[response.requestId](
+            new ResultSet(response.result.data, response.status.attributes)
```
Maybe instead of having 4 resultsets, the user wants to access each individual item.

tkolanko replied:
That would mean different parsing would have to take place between partial content and final parsing/if the total results were less than the batch size.

jorgebay reviewed (commented):
Good idea, I think the js api needs a way to access the results as they come.

I've commented mostly on the API below.

spmallette commented:
@tkolanko i just thought i'd follow up on this PR which seems to pass CI builds now. is this done from your perspective?

tkolanko commented:
> @tkolanko i just thought i'd follow up on this PR which seems to pass CI builds now. is this done from your perspective?

Yeah, good on my side if @jorgebay is good with it

jorgebay commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/client.js:
```diff
@@ -73,41 +72,118 @@ class Client {
-    if (this._options.session && this._options.processor === 'session') {
-      args['session'] = this._options.session;
+    if (this._options.session && this._options.processor === "session") {
+      args["session"] = this._options.session;
     }
 
     if (message instanceof Bytecode) {
-      if (this._options.session && this._options.processor === 'session') {
-        return this._connection.submit('session', 'bytecode', args, requestIdOverride);
+      if (this._options.session && this._options.processor === "session") {
+        return this._connection.stream(
+          "session",
+          "bytecode",
+          args,
+          requestIdOverride
+        );
       } else {
-        return this._connection.submit('traversal', 'bytecode', args, requestIdOverride);
+        return this._connection.stream(
+          "traversal",
+          "bytecode",
+          args,
+          requestIdOverride
+        );
       }
-    } else if (typeof message === 'string') {
-      args['bindings'] = bindings;
-      args['language'] = 'gremlin-groovy';
-      args['accept'] = this._connection.mimeType;
-      return this._connection.submit(this._options.processor || '','eval', args, requestIdOverride);
+    } else if (typeof message === "string") {
+      args["bindings"] = bindings;
+      args["language"] = "gremlin-groovy";
+      args["accept"] = this._connection.mimeType;
+      return this._connection.stream(
+        this._options.processor || "",
+        "eval",
+        args,
+        requestIdOverride
+      );
```
this code block is identical to the previous one, except from the underlying method that it's called.

Would it be possible to extract to something like `executeOnConnection()` or something like that, where the connection method is a parameter as well?

For example, it will have:
```javascript
return this._connection.[methodName](/* ... */)
```

tkolanko replied:
That's what I originally did in https://github.com/apache/tinkerpop/pull/1539/commits/fe1e468945ea3a36c53c2b452627bfffd20e0239 but it ended up breaking all the client tests that overwrote the internal connection object and I couldn't figure out why. I'll take another look to see if I can get the tests to pass with that

jorgebay replied:
It's just for maintainability but if you see that's too much work, we can move forward without extracting a common method.

jorgebay reviewed (commented):
Looks great, I've added some minor final suggestions below.

The formatting changes made reviewing a little harder but the logic is 👌

tkolanko commented on gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/client.js:
```diff
@@ -46,13 +45,13 @@ class Client {
   constructor(url, options = {}) {
     this._options = options;
-    if (this._options.processor === 'session') {
+    if (this._options.processor === "session") {
```
Yeah, it's most likely my editor. I'll get these fixed up

Would you be willing to accept a seperate pr that enabled eslint with airbnb config + prettier and just accept it's formatting?

jorgebay replied:
Sure, we feel free to create a JIRA ticket.
My main concern is flattening the history too much, so we can start with eslint w/ few presets, it's fine by me.

tkolanko commented:
@spmallette I think this will need to wait until after 3.5.3 so I can make the changes @jorgebay requested

tkolanko commented:
@spmallette @jorgebay Here's where we are with this PR:

1. I've fixed up the formatting issues and will create a new JIRA ticket to implement a standard for linting the JS files
2. The tests are passing, however, there is a duplicated block of code that should be extracted out. The current client tests are overriding an internal object and they did not play nice with having that block of code extracted out. I need more time to figure out why that is. We can either hold off on this PR and fixed that up or we can merge this PR in and I can create a JIRA ticket to figure out a better way to remove the duplicated block of code

jorgebay reviewed (approved):
This is a huge step forward in terms of user facing API. Thanks for looking into the proposed changes.

VOTE +1

jorgebay commented:
I'll leave it open for a few more hours, to give the opportunity for other committers to chime in.

If there aren't any remarks by then, I'll merge it.
